File: Arrays and Strings/1.7-Rotate-Matrix/Rotate.java
import java.util.*;
import java.io.*;

public class Rotate {

    public boolean rotate(int[][] M) {
	if (M.length == 0 || M[0].length != M.length) return false;

	for (int track = 0; track < M.length / 2; ++track) {
	    int begin = track;
	    int end = M.length - begin - 1;
	    for (int i = begin; i < end; ++i) {
		int diff = i - begin;
		int temp = M[track][i];
		M[track][i] = M[end - diff][begin];
		M[end - diff][begin] = M[end][end - diff];
		M[end][end - diff] = M[i][end];
		M[i][end] = temp;
	    }
	}

	return true;
    }

    public static void printMatrix(int[][] M) {
	for (int[] v : M) {
	    for (int x : v) {
		System.out.print(x + " ");
	    }
	    System.out.println();
	}
    }

    public static int[][] readMatrix(InputStream stream) {
	Scanner in = new Scanner(stream);
	ArrayList<ArrayList<Integer>> rows = new ArrayList<ArrayList<Integer>>();
	while (in.hasNext()) {
	    ArrayList<Integer> row = new ArrayList<Integer>();
	    String line = in.nextLine();
	    for (String s : line.split("\\s+")) {
		row.add(Integer.parseInt(s));
	    }
	    rows.add(row);
	}

	int[][] M = new int[rows.size()][];
	int i = 0;
	for (ArrayList<Integer> row : rows) {
	    int[] data = new int[row.size()];
	    for (int j = 0; j < data.length; ++j) {
		data[j] = row.get(j);
	    }
	    M[i++] = data;
	}

	return M;
    }

    public static void main(String[] args) {
	System.out.println("Enter an N x N matrix row by row (send EOF to end input): ");
	int[][] M = readMatrix(System.in);
	System.out.println("\nRotated:");
	if ((new Rotate()).rotate(M)) {
	    printMatrix(M);
	} else {
	    System.out.println("Invalid matrix dimensions.");
	}
    }
}

File: Arrays and Strings/1.7-Rotate-Matrix/rotate.cpp
#include <iostream>
#include <sstream>
#include <vector>

bool rotate(std::vector<std::vector<int>>& M) {
  /* Make sure the matrix is square */
  if (M.size() == 0 || M.size() != M[0].size()) return false;

  /* We will rotate track by track (like a disk) starting with
     the outer track */
  for (int track = 0; track < M.size() / 2; ++track) {
    int begin = track;
    int end = M.size() - begin - 1;
    for (int i = begin; i < end; ++i) {
      int diff = i - begin;
      /* Temporarily store the value at the top */
      int temp = M[begin][i];
      /* Move from left to top */
      M[begin][i] = M[end - diff][begin];
      /* Move from bottom to left */
      M[end - diff][begin] = M[end][end - diff];
      /* Move from right to bottom */
      M[end][end - diff] = M[i][end];
      /* Move from temp (top) to right */
      M[i][end] = temp;
    }
  }
  /* Matrix was successfully rotated */
  return true;
}

void printMatrix(std::vector<std::vector<int>> M) {
  for (std::vector<int> v : M) {
    for (int x : v) {
      std::cout << x << " ";
    }
    std::cout << std::endl;
  }
}

std::vector<std::vector<int>> readMatrix(std::istream& in) {
  std::vector<std::vector<int>> M;
  std::string line;
  while (getline(in, line)) {
    std::stringstream ls(line);
    int x;
    std::vector<int> v;
    while (ls >> x) {
      v.push_back(x);
    }
    M.push_back(v);
  }
  return M;
}



int main() {
  std::cout << "Enter an N x N matrix row by row (send EOF to end input): " << std::endl;
  std::vector<std::vector<int>> M = readMatrix(std::cin);
  std::cout << std::endl << "Rotated: " << std::endl;
  if (rotate(M)) printMatrix(M);
  else std::cout << "Invalid matrix dimensions." << std::endl;
  return 0;
}
